### PR TITLE
Initial support for CodeRefactorings ordering.

### PR DESCRIPTION
**Pull request by petrroll**
Work for: #35525

### CODE REVIEW

**Comment by petrroll** [comment]
Please take a look if this approach is reasonable. 
Will add comments & cleanup names if it is.

File: src/Features/Core/Portable/UseNamedArguments/AbstractUseNamedArgumentsCodeRefactoringProvider.cs [hunk 1]
```diff
@@ -30,7 +30,8 @@ protected abstract class Analyzer<TBaseArgumentSyntax, TSimpleArgumentSyntax, TA
             {
                 var (document, textSpan, cancellationToken) = context;
 
-                var argument = await context.TryGetRelevantNodeAsync<TBaseArgumentSyntax>().ConfigureAwait(false) as TSimpleArgumentSyntax;
+                var potentialArguments = await document.GetRelevantNodesAsync<TBaseArgumentSyntax>(textSpan, cancellationToken).ConfigureAwait(false);
+                var argument = potentialArguments.FirstOrDefault(n => n?.Parent is TArgumentListSyntax) as TSimpleArgumentSyntax;
```
Comment (petrroll) [inline src/Features/Core/Portable/UseNamedArguments/AbstractUseNamedArgumentsCodeRefactoringProvider.cs]: This is ordering unrelated change made to surface the refactoring in more situation & to help showcase better ordering.

File: src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs [hunk 2]
```diff
@@ -52,8 +52,11 @@ public override async Task ComputeRefactoringsAsync(CodeRefactoringContext conte
 
             var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
 
-            context.RegisterRefactoring(new MyCodeAction(CSharpFeaturesResources.Convert_to_method,
-                c => UpdateDocumentAsync(root, document, parentBlock, localFunction, c)));
+            context.RegisterRefactoring(
+                new MyCodeAction(CSharpFeaturesResources.Convert_to_method,
+                            c => UpdateDocumentAsync(root, document, parentBlock, localFunction, c)),
+                localFunction.Span
+                );
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs]: ehhh.... :)

**Comment by CyrusNajmabadi** [comment]
Will wait until you want eyes.

**Comment by petrroll** [comment]
I could use review of the overall architecture. There're many ways to do it am not sure if this one is necessarily the best.

**Comment by petrroll** [comment]
Also considered an alternative where instead of changing `.CodeActions` to `<(CodeAction,TextSpan)>` another `ImmutableArray<TextSpan> ApplicableToSpan` was added but since the elements are logically linked a tuple solution seems more appropriate.

File: src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs (hunk 2 above)
Comment (petrroll) [reply src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs]: Yeah, I need to figure out better wrapping. That said, they all look terrible.

Comment (sharwell) [reply src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs]:
```
context.RegisterRefactoring(
  new MyCodeAction(
    CSharpFeaturesResources.Convert_to_method,
    c => UpdateDocumentAsync(root, document, parentBlock, localFunction, c)),
  localFunction.Span);
```

File: src/Features/Core/Portable/CodeRefactorings/CodeRefactoring.cs [hunk 3]
```diff
@@ -16,14 +17,14 @@ internal class CodeRefactoring
         /// <summary>
         /// List of possible actions that can be used to transform the code.
         /// </summary>
-        public ImmutableArray<CodeAction> Actions { get; }
+        public ImmutableArray<(CodeAction action, TextSpan? applicableToSpan)> CodeActions { get; }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeRefactorings/CodeRefactoring.cs]: so i can see this is the crux of hte change.  You're passing along this additional piece of info.  Can you give me a rundown (or point me to where you've already done this) on what this piece of data will be used for?  What's the intuition behind how it will change the user experience?  Thanks!

Comment (petrroll) [reply src/Features/Core/Portable/CodeRefactorings/CodeRefactoring.cs]: Long version: https://github.com/dotnet/roslyn/issues/35525#issuecomment-517046330

Short version: AoRN the `CodeRefactoring` & `SuggestedActionsSource` services sort all retrieved (code-fixes, refactorings) by (among others) `ApplicableToSpan`. The `ApplicableToSpan` for (all) Refactorings is, however, set to current selection.

This change enables Refactorings to [register](https://github.com/dotnet/roslyn/pull/37777/files#diff-4b18194545a98c011598321cbfa1dd9fR67) their `ApplicableToSpan` and [flows](https://github.com/dotnet/roslyn/pull/37777/files#diff-2395e5c70cda2d55f3f315c63f3cc59dR789) it to the `SuggestedActionsSource ` to be used for sorting (instead of [statically setting](https://github.com/dotnet/roslyn/pull/37777/files#diff-2395e5c70cda2d55f3f315c63f3cc59dR760) the current selection)

The UX change is that we will be able to order refactorings with respect to how far / close they're to current selection instead of relying on static ordering by an attribute.

E.g. `foo((bar($$), 2)` will show add arg. name (for `bar`) first, add tuple ele. name (for `bar()` second; while `foo(($$2, 2))` will show the one for tuple first and for `foo`'s parameter second.

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs [hunk 4]
```diff
@@ -776,28 +776,28 @@ private static SuggestedActionSetPriority GetSuggestedActionSetPriority(CodeActi
             {
```
Comment (petrroll) [inline src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs]: The applicableSpan parameter is obsolete, will remove it on cleanup.

File: src/Features/Core/Portable/CodeRefactorings/CodeRefactoring.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/CodeRefactorings/CodeRefactoring.cs]: Thanks.  I'll read through that and review later :)

Comment (petrroll) [reply src/Features/Core/Portable/CodeRefactorings/CodeRefactoring.cs]: Sure, as I said in comment above, there're more ways to route the information so I'm looking for architectural feedback mostly.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/CodeRefactorings/CodeRefactoring.cs]: I'd probably just use a named struct for this.  But overall i have no issue with this :)

File: src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs [hunk 5]
```diff
@@ -48,14 +62,16 @@ public struct CodeRefactoringContext
         /// Add supplied <paramref name="action"/> to the list of refactorings that will be offered to the user.
         /// </summary>
         /// <param name="action">The <see cref="CodeAction"/> that will be invoked to apply the refactoring.</param>
-        public void RegisterRefactoring(CodeAction action)
+        public void RegisterRefactoring(CodeAction action) => RegisterRefactoring(action, null);
+
+        internal void RegisterRefactoring(CodeAction action, TextSpan? applicableToSpan)
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: This should eventually be made a public API, whenever you are ready. Also why is the text span a nullable value type parameter? Callers should always pass in a valid applicable span OR use the overload without the span. You can probably have a shared private method that has signature and is invoked from both of them

We should also write an analyzer at https://github.com/dotnet/roslyn-analyzers/tree/master/src/Microsoft.CodeAnalysis.Analyzers/Core/FixAnalyzers that flags use of the prior overload without span if the second overload is available. Can you file a feature request issue on roslyn-analyzers repo for this?

File: src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs [hunk 6]
```diff
@@ -48,14 +62,16 @@ public struct CodeRefactoringContext
         /// Add supplied <paramref name="action"/> to the list of refactorings that will be offered to the user.
         /// </summary>
         /// <param name="action">The <see cref="CodeAction"/> that will be invoked to apply the refactoring.</param>
-        public void RegisterRefactoring(CodeAction action)
+        public void RegisterRefactoring(CodeAction action) => RegisterRefactoring(action, null);
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: so this.Span is the selection or caret position right? which currently what engine uses for applicable span
http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures.Wpf/Suggestions/SuggestedActionsSource.cs,672

so here. I would do something like RegisterRefactoring(action, Span) which will retain same behavior as before. and mark this API [Obsolete] and make the other one public.

give Span here let you remove ? and remove any special case on engine side.

(action, span) =>

File: src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs (hunk 5 above)
Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: The reason for using nullable was that this channel needs to be able to carry CodeActions without the `TextSpan` (i.e. when the original overload is used) as well. There's nothing preventing us from representing missing `TextSpan` as `default` instead of `null` but thought `null` was more clear.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: I think the point is, you should have these signatures:

```
RegisterRefactoring(CodeAction)
RegisterRefactoring(CodeAction, TextSpan)
```

INternally, these can sit on top of a private:

```RegisterRefactoring(CodeAction, TextSpan?)```

Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: Ah, I see. That's doable but the nullability would still leak via the constructor and the saved lambda and it generally doesn't end up very clean. 

Representing missing `applicableToSpan` as default is an overall better solution.

**Comment by petrroll** [comment]
TODO: 
- Move more refactorings to `RegisterRefactoring(...TextSpan)`
- Either obsolete old `RegisterRefactoring` or request analyzer.
- Improve `applicableToComparer`

File: src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs (hunk 6 above)
Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: I'm not sure I follow the first question the `TextSpan` here is the span where the CodeAction (Refactoring) should be applicable. If it doesn't get populated here then a `default(TextSpan)` is used for the Action's `applicableToSpan` which might have small ordering implications but nothing too drastic in comparison to what happens today (static ordering by attribute).

As for the ordering engine, changes there will come soon. 
- Improve the comparer for `applicableToSpan` so it uses selection info properly 
- Potentially remove the difference in handling refactorings and codeFixes (as you're saying).

File: src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: > Ah, I see. That's doable but the nullability would still leak via the constructor

Why?

Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: You're right, wrong assumption. Fixed.

**Comment by CyrusNajmabadi** [comment]
Will try to get to this today.  If i don't, feel free to ping me tomorrow.

File: src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs [hunk 7]
```diff
@@ -48,14 +65,25 @@ public struct CodeRefactoringContext
         /// Add supplied <paramref name="action"/> to the list of refactorings that will be offered to the user.
         /// </summary>
         /// <param name="action">The <see cref="CodeAction"/> that will be invoked to apply the refactoring.</param>
-        public void RegisterRefactoring(CodeAction action)
+        public void RegisterRefactoring(CodeAction action) => RegisterRefactoring(action, applicableToSpan: default);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: this seems... 'wrong' to me.  If there is no span provided by the actual refactoring, would we not just default to this.Span here?  because, effectively, the refactoring is def saying "i apply to the span you passed in" :)   

Isn't the optional parameter only if the refactoring wants to override that with something more specific from context?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: Also, if we can manage to do this, then we'll be in the state where there's no need for nullable-spans at all.  Because we'll always have a valid span to consider.

**Comment by CyrusNajmabadi** [comment]
Open design question i'd like thought on before proceeding with the review.

File: src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs (hunk 7 above)
Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: Note: 
- `RegisterRefactoring(CodeAction action)` shouldn't pass default but `null` (oversight after switching to `TextSpan?`)

I somehow absolutely missed that there's `this.Span` on that type. I'll need to check what it gets set to (I actually have no idea).

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: "this.Span" is literally the span the user has selected (or hte position the caret is at if there's no selection).  It's what all refactorings *start* with to determine if they should show :)  i.e. in all ComputeRefactoringsAsync things generally start with:

```c#
        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
        {
            var (document, span, cancellationToken) = context;
```

`this.Span` is the span the refactoring starts examining.  So, in general, if the refactoring itself doesn't report a better 'applicabilitySpan' it seems reasonable to start with the one we know it was examining.

Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: Ah, right, somehow I didn't realize this is the same Context class (long day :)). 

As for using that, that would cause these refactorings (the ones who didn't opt-in) always closest while I somehow feel we might want to do the very opposite / keep them in their original ordering by the attribute (which is what happens if they specify null right now).

File: src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs (hunk 6 above)
Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: I understand the question about the span now :). It's being discussed below.

File: src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: you could have a bool if it was the 'default' span that came along, or it was explicitly provided, and order accordingly :)   i'm honestly not sure what the best thing is here.  We'll likely have to try out a few things to see what feels right :)

Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: TBH it doesn't really matter much. 

We actually can't pass null properly without leaking the `TextSpan?` as part of public API. The registration lambda gets passed as part in (public) constructor and it needs to be accept whatever we want to put to our (internal) `ImmutableArray<…> CodeActions` storage.

But that means we just can't call the lambda with `null` in the original `RegisterRefactoring()` overload.

And having constructor that takes register lambda accepting `TextSpan?` but having `RegisterRefactoring` method that can only take `TextSpan` is just weird.

Thus, I'd just go with passing default to make them "furthest" instead of `this.Span` and thus making them closest.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: > We actually can't pass null properly without leaking the TextSpan? as part of public API. The registration lambda gets passed as part in (public) constructor and it needs to be accept whatever we want to put to our (internal) ImmutableArray<…> CodeActions storage.

Can you point to what you mean?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: > Thus, I'd just go with passing default

This doesn't work in the case of refactorings at the start of a fiel :) (which do exist) :)

Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: > This doesn't work in the case of refactorings at the start of a fiel :) (which do exist) :)

I don't follow.

> Can you point to what you mean?

Imagine [this](https://github.com/dotnet/roslyn/pull/37777/commits/b82e2c2baa57fe50af62353e3a1470ca915dba2b#diff-a2c30b83123c3cb46ac84dae8175bc4bR24) accepts null and we want to pass `null` [here](https://github.com/dotnet/roslyn/pull/37777/commits/b82e2c2baa57fe50af62353e3a1470ca915dba2b#diff-4b18194545a98c011598321cbfa1dd9fR68) (carries info "didn't register specific TextSpan") .  The `RegisterRefactoring` [just calls](https://github.com/dotnet/roslyn/pull/37777/commits/b82e2c2baa57fe50af62353e3a1470ca915dba2b#diff-4b18194545a98c011598321cbfa1dd9fL86) `_registerRefactoring` lambda with given parameters and so if we want to call it with `null` the [lambda](https://github.com/dotnet/roslyn/pull/37777/commits/b82e2c2baa57fe50af62353e3a1470ca915dba2b#diff-4b18194545a98c011598321cbfa1dd9fR30) needs to be of type `Action<CodeAction, TextSpan?>` (has to accept nullable `TextSpan`).

The lambda, however, isn't an implementation detail. It gets set in [*public* constructor](https://github.com/dotnet/roslyn/pull/37777/commits/b82e2c2baa57fe50af62353e3a1470ca915dba2b#diff-4b18194545a98c011598321cbfa1dd9fR52). And that leaks the information that the registration should somehow work with Nullable `applicableToSpan` 🤷‍♂️.

It's not terrible, nobody outside probably uses the constructor. But it _just feels weird_. Either have it nullable explicitly in `RegisterRefactoring` (but that communicates wrong idea about it being optional) or have it nowhere.

> Note: we [still support](https://github.com/dotnet/roslyn/pull/37777/commits/b82e2c2baa57fe50af62353e3a1470ca915dba2b#diff-4b18194545a98c011598321cbfa1dd9fR45) the original constructor without `TextSpan applicableToSpan` parameter

Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: Generally, going with `default` is wrong because it doesn't even have to be the furthest (it's a valid `TextSpan(0, 0)`, right? But it's not wrong enough for it to matter. The worst case scenario is we'll order something slightly lower/higher than we should which IMO isn't all that consequential. Especially given we haven't even order things at all right now.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: > I don't follow.

Passing default is just passing the `[0,0)` span.  But that span has meaning, and it would be indistinguishable from someone actually truly passing that span explicitly.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: > The lambda, however, isn't an implementation detail. It gets set in public constructor. 

?

From your own link: `internal CodeRefactoringContext(`

It's not a public constructor.  It's internal.  We can do whatever we want with it, up to and including making it take a `TextSpan?` or making it take a `TextSpan+bool`.

Basically, it seems to me that we have 100% control of the end to end here.  So we can just pass along the information as accurately as possible.  This can then be used by the end suggestion source in whatever way it sees fit.

So, personally, i would the information to pass should be:

1. the code action.
2. the applicable span for the code action (either explicitly provided by user code, or implicitly provided by the context).
3. a bool to say if this was explicitly provided or not.

This change can be made entirely in the internal surface area of this library.

The final suggestion source can then use this as it sees fit to order appropriately.  There is no ambiguity.  No overlapping implicit and explicit values.  etc. etc. 

Thoughts?

Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: > From your own link: internal CodeRefactoringContext(

You're absolutely right. I don't know why I was _convinced_ it's public struct. I'm for nullability instead of <bool, TextSpan>.

> Passing default is just passing the [0,0) span. But that span has meaning, and it would be indistinguishable from someone actually truly passing that span explicitly.

Ah, yep as said in the comment just above I didn't consider that being such an issue. 

Will update it once more (let's hope for the last time this time :D )

Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: Wait, it's internal for now but it will (hopefully) be exposed to be public soon. It's a public struct.

Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: THB the more I think about it, the more I'm leaning towards going the `TextSpan?` route and the fact that the lambda in (eventually *public*) constructor will accept `TextSpan` isn't all that bad leakage of implementation detail.

Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: Thoughts @CyrusNajmabadi @heejaechang ?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: Don't do this.  The intent is that hte structs are temporary context objects created by the engine.  It's not intended that they be publicly constructable.  this is a super common pattern we use all over the place *including* for public APIs.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: > Just as the RegisterCodeAction overload for TextSpan is currently internal only for now but will (hopefully) be made public (once we all agree it's stable, that's why I want to do it in two stages).

Making this new method public in the future is fine.  It's expected that 3rd parties would use it.  That's different from the context-struct object.  These are *often* *intentionally* internal as they serve to bridge between the context object and the owning internal coordinator object.  Users will never create these themselves, and things like the lambda/delegate are 100% internal implementation details we would never expose to them.

Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: But it already is publicly creatable: http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/CodeRefactorings/CodeRefactoringContext.cs,35

And having a public constructor that is obsolete and then internal one that people should use instead (but can't) doesn't make sense.

Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: Sure, the struct should've been an implementation detail not construtable by user but it is and we have to deal with it somehow conistently.

Or am I missing something (I feel like I might be)?

Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: > like the lambda/delegate are 100% internal implementation details we would never expose to them. 

That's the issue, it has already happened. 

> Just a nitpic: GitHub has serious delay between you writing the message & me seeing it, hence why my responses are sort of weird (as if I haven't read what you wrote)

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: *son of a*!

Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: You see the problem now and while all of my proposed solutions are sub-optimal?

Comment (petrroll) [reply src/Workspaces/Core/Portable/CodeRefactorings/CodeRefactoringContext.cs]: Resolution from gitter: 
 > there's a strong arg to be made for just obsoleting those guys and just keeping the new guys internal

> we have precedence for this.

**Comment by petrroll** [comment]
Note: I seem to be blocked on an integration test I don't know how to debug (can't manage run it locally), manually checking the source it checks CodeActions for seem identical to previous VS so I can't see what I broke.

TODO:
- Fix the broken test (will need help)
- Update TextSpan to be nullable again & remember to not-expose the new Constructor & obsolete the old.

**Comment by petrroll** [comment]
Ready to me merged, will introduce actually passing the TextSpan in next PR.

**Comment by petrroll** [comment]
cc @CyrusNajmabadi is the comparer better explained here?

**Comment by CyrusNajmabadi** [comment]
> cc @CyrusNajmabadi is the comparer better explained here?

AWESOME.  THanks :)

**Comment by petrroll** [comment]
Superseded by reviewed #37941.
